src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs

```diff
@@ -1202,6 +1213,146 @@ private void EmitSwitchDispatch(BoundSwitchDispatch dispatch)
+            }
+
+            void emitCharDispatches(LengthBasedStringSwitchData lengthBasedSwitchInfo, LocalOrParameter keyTemp, LabelSymbol fallThroughLabel, SyntaxNode syntaxNode)
+            {
+                var charType = Binder.GetSpecialType(_module.Compilation, SpecialType.System_Char, syntaxNode, _diagnostics);
+                var charTemp = AllocateTemp(charType, syntaxNode);
+
+                foreach (var charJumpTable in lengthBasedSwitchInfo._charJumpTables)
+                {
+                    // label for CharJumpTable:
+                    _builder.MarkLabel(charJumpTable.label);
+
+                    //   charTemp = keyTemp[selectedCharPosition];
+                    if (isSpanOrReadOnlySpan)
+                    {
+                        _builder.EmitLoadAddress(keyTemp);
+                    }
+                    else
+                    {
+                        _builder.EmitLoad(keyTemp);
+                    }
+                    _builder.EmitIntConstant(charJumpTable.selectedCharPosition);
+                    _builder.EmitOpCode(ILOpCode.Call, stackAdjustment: -1);
+                    emitMethodRef(indexerRef);
+                    if (isSpanOrReadOnlySpan)
+                    {
+                        _builder.EmitOpCode(ILOpCode.Ldind_u2);
+                    }
+                    _builder.EmitLocalStore(charTemp);
+
+                    // switch dispatch on charTemp using fallThroughLabel and cases:
+                    //   charConstant -> corresponding label
+                    _builder.EmitIntegerSwitchJumpTable(
+                        charJumpTable.charCaseLabels.Select(p => new KeyValuePair<ConstantValue, object>(p.value, p.label)).ToArray(),
+                        fallThroughLabel, charTemp, charType.PrimitiveTypeCode);
+                }
+
+                FreeTemp(charTemp);
+            }
+
+            void emitFinalDispatches(LengthBasedStringSwitchData lengthBasedSwitchInfo, LocalOrParameter keyTemp, TypeSymbol keyType, LabelSymbol fallThroughLabel, SyntaxNode syntaxNode)
+            {
+                foreach (var stringJumpTable in lengthBasedSwitchInfo._stringJumpTables)
+                {
+                    // label for StringJumpTable:
+                    _builder.MarkLabel(stringJumpTable.label);
+
+                    // switch dispatch on keyTemp using fallThroughLabel and cases:
+                    //   stringConstant -> corresponding label
+                    EmitStringSwitchJumpTable(
+                        stringJumpTable.stringCaseLabels.Select(p => new KeyValuePair<ConstantValue, object>(p.value, p.label)).ToArray(),
+                        fallThroughLabel, keyTemp, syntaxNode, keyType);
+                }
+            }
+
+            void emitMethodRef(Microsoft.Cci.IMethodReference lengthMethodRef)
+            {
+                var diag = DiagnosticBag.GetInstance();
+                _builder.EmitToken(lengthMethodRef, syntaxNode: null, diag);
```

Comment:
We also don't use the real diagnostic bag for existing logic emitting length method (line 1468). 
The length method is checked during binding/lowering. That's why we don't get here if there's a problem with the method.
On the other hand, I suppose that since there are not diagnostics, it wouldn't hurt to use the common bag... I'm leaning to stick with precedent for this one (assertion tightens things a tiny bit).